provider/pkg/gen/examples.go

```
@@ -312,44 +312,43 @@ func generateExamplePrograms(example resources.AzureAPIExample, body *model.Body
 	return languageExample, nil
 }
 
-// Panic after 3 minutes, unless cancel is called.
-func makeTimeout(timer string) (cancel func()) {
-	timeout := make(chan struct{})
-	go func() {
-		for {
-			select {
-			case <-timeout:
-				return
-			case <-time.After(3 * time.Minute):
-				log.Printf("%s timed out", timer)
-				panic("timeout")
-			}
-		}
-	}()
-	return func() { timeout <- struct{}{} }
-}
-
 func recoverableProgramGen(name string, program *hcl2.Program, fn programGenFn) (files map[string][]byte, err error) {
 	defer func() {
 		if r := recover(); r != nil {
 			err = fmt.Errorf("panic recovered during generation: %v", r)
 		}
 	}()
 
-	var d hcl.Diagnostics
-	cancel := makeTimeout(name)
-	defer cancel()
-	files, d, err = fn(program)
-
-	if err != nil {
-		return nil, err
+	type programGenResult struct {
+		files map[string][]byte
+		d     hcl.Diagnostics
+		err   error
 	}
-	if d.HasErrors() {
-		err := program.NewDiagnosticWriter(os.Stderr, 0, true).WriteDiagnostics(d)
-		if err != nil {
-			log.Printf("failed to write diagnostics: %v", err)
+
+	c := make(chan programGenResult, 1)
+	go func() {
+		var d hcl.Diagnostics
+		files, d, err = fn(program)
+		c <- programGenResult{files, d, err}
+	}()
+
+	select {
+	case res := <-c:
+		if res.err != nil {
+			return nil, res.err
 		}
+		files = res.files
+		if res.d.HasErrors() {
+			err = program.NewDiagnosticWriter(os.Stderr, 0, true).WriteDiagnostics(res.d)
+			if err != nil {
+				log.Printf("failed to write diagnostics: %v", err)
+			}
+		}
+	case <-time.After(3 * time.Minute):
+		log.Printf("%s timed out after 3m", name)
```

On a second pass, do we actually want to return an error if it timed out?

---

you're right, good catch